hydra-node/test/Hydra/Chain/Direct/TxSpec.hs

```diff
@@ -1,24 +1,87 @@
-import Test.QuickCheck.Hedgehog (hedgehog)
+import Cardano.Binary (serialize)
+import Cardano.Ledger.Alonzo.Data (Data (Data))
+import Cardano.Ledger.Alonzo.Tx (ValidatedTx (ValidatedTx, wits, body), outputs)
+import Cardano.Ledger.Alonzo.TxWitness (TxWitness (txdats), nullDats, unTxDats)
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Map as Map
+import Hydra.Chain (HeadParameters (..), toOnChainTx, PostChainTx (InitTx))
+import Hydra.Chain.Direct.Tx (initTx, observeTx, constructTx)
+import Hydra.Contract.ContestationPeriod (contestationPeriodFromDiffTime)
+import Hydra.Contract.Head (State (Initial))
+import Hydra.Contract.Party (partyFromVerKey)
+import Hydra.Party (vkey)
+import Plutus.V1.Ledger.Api (toBuiltinData, toData)
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import Test.QuickCheck (counterexample, (===), (==>))
+import Hydra.Ledger.Simple (SimpleTx)
+import Hydra.Chain.Direct (Era)
+import Cardano.Ledger.Mary.Value (Value(Value), PolicyID, AssetName)
+import Cardano.Ledger.Alonzo (TxOut)
+import Cardano.Ledger.Alonzo.TxBody (TxOut(TxOut))
+import Cardano.Ledger.Crypto (StandardCrypto)
 
 spec :: Spec
 spec =
-  parallel $
+  parallel $ do
+    prop "observeTx . constructTx roundtrip" $ \postTx txIn time ->
+      isImplemented postTx ==> -- TODO(SN): test all constructors
+      observeTx (constructTx txIn postTx) === Just (toOnChainTx @SimpleTx time postTx)
+
     describe "initTx" $ do
-      prop "can be used to construct a non-balanced TxBody" $ \params ->
-        forAll (hedgehog genTxIn) $ \txIn ->
-          case makeTransactionBody $ initTx params txIn of
-            Left err -> counterexample ("TxBodyError: " <> show err) False
-            Right _ -> property True
+      prop "can construct & serialize unsigned initTx" $ \txIn params ->
+        let tx = initTx params txIn
+            cbor = serialize tx
+            len = LBS.length cbor
+         in counterexample ("Tx: " <> show tx) $
+              counterexample ("Tx serialized size: " <> show len) $
+                len < 16000 -- TODO(SN): use real max tx size
+
+      prop "contains some datums" $ \txIn params ->
+        let ValidatedTx{wits} = initTx params txIn
+            dats = txdats wits
+         in counterexample ("TxDats: " <> show dats) $
+              not $ nullDats dats
+
+      prop "contains HeadParameters as datums" $ \txIn params ->
+        let ValidatedTx{wits} = initTx params txIn
+            dats = txdats wits
+            HeadParameters{contestationPeriod, parties} = params
+            onChainPeriod = contestationPeriodFromDiffTime contestationPeriod
+            onChainParties = map (partyFromVerKey . vkey) parties
+            datum = Initial onChainPeriod onChainParties
+         in Map.elems (unTxDats dats) === [Data . toData $ toBuiltinData datum]
+
+      -- TODO(SN): assert monetary policy?
+      prop "distributes participation tokens (expected failure)" $ \txIn params ->
```

Comment:
YAGNI? or `pendingWith` a string rather than a comment and `True` property

Reply:
There is no `pendingWith` in Quickcheck right? I was leaving this in for later and we can of course remove it altogether before merging this.